File: resolveinc.pl
#!/usr/bin/perl
# Use on onelua.c to compile it into a single file.

use strict;
use warnings;

use File::Slurp;
use File::Basename qw(dirname basename);

my $file = $ARGV[0];

my %verboten = map { $_ => 1 } qw(lua.h lualib.h lauxlib.h);

sub process_file {
  my ($file, $hash, $depth) = @_;
  open(my $fh, "<", $file) or die "can't open file $file: $!";
  my $dir = dirname($file);
  my $open = 0;
  while (my $line = <$fh>) {
    $line =~ s/\/\*.*?\*\///g;
    if ($open && $line =~ m/^.*?\*\//) {
      $line =~ s/^.*?\*\///;
      $open = 0;
    }
    if (!$open && $line =~ m/\/\*.*?$/) {
      $open = 1;
      $line =~ s/\/\*.*?$//;
      print $line . "\n" if $line !~ m/^\s*$/;
    }
    if (!$open) {
      if ($line =~ m/^#include\s+"(\S+)"/ && !$verboten{$1}) {
        if (!$hash->{$1} && (-f $dir . "/" . $1)) {
          $hash->{$1} = 1;
          process_file($dir . "/" . $1, $hash, $depth + 1);
        }
      } elsif ($line !~ m/^\s*$/) {
        print $line;
      }
    }
  }
}

print "#define MAKE_LIB\n" if $file =~ m/onelua\.c/i;;
print "#define LUA_CORE
#define LUA_LIB
#define ltable_c
#define lvm_c\n" if $file =~ m/lua\.h/i;
process_file($file, { }, 0);

File: utilities/resolveinc.lua
#!/usr/bin/env lua5.4
-- Use on onelua.c to compile it into a single file.
-- Primarily used to for wtk itself.

local verboten = { ["lua.h"] = 1, ["lualib.h"] = 1, ["lauxlib.h"] = 1 }
local processed = {}

local file = assert(..., "requires a file to be specified")

local function process_file(file, depth)
  local open = false
  local dir = file:match("^(.+)/")
  for line in io.lines(file) do
    line = line:gsub("/%*.-%*/", "")
    if open and line:find("^.-%*/") then
      line = line:gsub("^.-%*/", "")
      open = false
    end
    if not open and line:find("/%*") then 
      open = true
      line = line:gsub("/%*.*$", "")
      if line:find("%S") then print(line .. "\n") end
    end
    if not open then
      local include = line:match("^%s*#include%s+\"(%S+)\"")
      if not open and (include and not verboten[include]) and io.open(dir .. "/" .. include) then 
        if not processed[include] then
          processed[include] = true
          process_file(dir .. "/" .. include, depth + 1)
        end
      elseif line:find("%S") then
        print(line)
      end
    end
  end
end

if file:match("onelua%.c") then
  print("#define MAKE_LIB")
elseif file:match("lua%.h") then
  print("#define LUA_CORE\n#define LUA_LIB\n#define ltable_c\n#define lvm_c\n#define WTK_BUNDLED_LUA")
end
process_file(file, 0);
